src/container/srv_target.c

```diff
@@ -279,38 +485,163 @@ ds_cont_child_cache_destroy(struct daos_lru_cache *cache)
+	} else {
+		D_ASSERT(!cont_child_started(cont_child));
+	}
+}
+
+void
+ds_cont_child_stop_all(struct ds_pool_child *pool_child)
+{
+	d_list_t		*cont_list;
+	struct ds_cont_child	*cont_child;
+
+	D_DEBUG(DF_DSMS, DF_UUID"[%d]: Stopping all containers\n",
+		DP_UUID(pool_child->spc_uuid),
+		dss_get_module_info()->dmi_tgt_id);
+
+	D_ASSERT(d_list_empty(&pool_child->spc_list));
+
+	cont_list = &pool_child->spc_cont_list;
+	while (!d_list_empty(cont_list)) {
+		cont_child = d_list_entry(cont_list->next,
+					  struct ds_cont_child, sc_link);
+		cont_child_stop(cont_child);
+	}
+}
+
+static int
+cont_child_start(struct ds_pool_child *pool_child, const uuid_t co_uuid,
+		 struct ds_cont_child **cont_out)
+{
+	struct dsm_tls		*tls = dsm_tls_get();
+	struct ds_cont_child	*cont_child;
+	int			 tgt_id = dss_get_module_info()->dmi_tgt_id;
+	int			 rc;
+
+	D_DEBUG(DF_DSMS, DF_CONT"[%d]: Starting container\n",
+		DP_CONT(pool_child->spc_uuid, co_uuid), tgt_id);
+
+	rc = cont_child_lookup(tls->dt_cont_cache, co_uuid,
+			       pool_child->spc_uuid, &cont_child);
+	if (rc) {
+		D_CDEBUG(rc != -DER_NONEXIST, DLOG_ERR, DF_DSMS,
+			 DF_CONT"[%d]: Load container error:%d\n",
+			 DP_CONT(pool_child->spc_uuid, co_uuid), tgt_id, rc);
+		return rc;
+	}
+
+	/*
+	 * The container is in stopping because:
+	 * 1. Container is going to be destroyed, or;
+	 * 2. Pool is going to be destroyed, or;
+	 * 3. Pool service is going to be stopped;
+	 */
+	if (cont_child->sc_stopping) {
+		D_ERROR(DF_CONT"[%d]: Container is in stopping\n",
+			DP_CONT(pool_child->spc_uuid, co_uuid), tgt_id);
+		rc = -DER_SHUTDOWN;
+	} else if (!cont_child_started(cont_child)) {
+		rc = cont_start_agg_ult(cont_child);
+		if (!rc) {
+			d_list_add_tail(&cont_child->sc_link,
+					&pool_child->spc_cont_list);
+			ds_cont_child_get(cont_child);
+		}
+	}
+
+	if (!rc && cont_out != NULL) {
+		*cont_out = cont_child;
+		ds_cont_child_get(cont_child);
+	}
+
+	/* Put the ref from cont_child_lookup() */
+	ds_cont_child_put(cont_child);
+	return rc;
+}
+
+static int
+cont_child_start_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		    vos_iter_type_t type, vos_iter_param_t *iter_param,
+		    void *data, unsigned *acts)
+{
+	struct ds_pool_child	*pool_child = data;
+
+	return cont_child_start(pool_child, entry->ie_couuid, NULL);
 }
 
-static void
-cont_child_put(struct daos_lru_cache *cache, struct ds_cont_child *cont)
+int
+ds_cont_child_start_all(struct ds_pool_child *pool_child)
 {
-	daos_lru_ref_release(cache, &cont->sc_list);
+	vos_iter_param_t	iter_param = { 0 };
+	struct vos_iter_anchors	anchors = { 0 };
+	int			rc;
+
+	D_DEBUG(DF_DSMS, DF_UUID"[%d]: Starting all containers\n",
+		DP_UUID(pool_child->spc_uuid),
+		dss_get_module_info()->dmi_tgt_id);
+
+	iter_param.ip_hdl = pool_child->spc_hdl;
+	/* The quantity of container is small, no need to yield */
+	rc = vos_iterate(&iter_param, VOS_ITER_COUUID, false, &anchors,
+			 cont_child_start_cb, (void *)pool_child);
```

Comment:
there is possibly with many many (several 100K?) containers in one pool, do we need to create ds_cont_child for each when pool connect? More over will create one aggregate ULT per container? (even that only very few container is with active operation and most are not opened).

Reply:
There will be at most hundreds of containers per pool, see https://github.com/daos-stack/daos/blob/master/doc/overview/storage.md

Reply:
Ah, ok, I was checking from an old pdf hld possibly too old.